### DISCUSSION TITLE
Source Generators is taking generated files as input when EmitCompilerGeneratedFiles and CompilerGeneratedFilesOutputPath are set

### DISCUSSION THREAD

**Discussion by papafe** [dotnet/roslyn#64705] [Ideas] [answered]
If in a project using source generators I want to emit the generated files to the file system I can add the following to the project file:
```
<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
<CompilerGeneratedFilesOutputPath>Generated</CompilerGeneratedFilesOutputPath>
```

This indeed emits the generated files in the filesystem in the `Generated` folder, but following runs of the source generator will also consider those generated files as part of files given as an input to the source generator again. This sounds definitely counterintuitive, as I would expect that all the files in `Generated` would not be run through the source generation again. 
It is possible to exclude the folder from compilation with `<Compile Remove="$(CompilerGeneratedFilesOutputPath)/**/*.cs" />`, but I would expect this to be the "default behaviour".

**Comment by PathogenDavid** [comment, marked as answer]
The default behavior is that [generated files go into the intermediate directory](https://github.com/dotnet/roslyn/blob/c01ad7600a0d4cef7196c0fd43bc0132bf0238cf/src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets#L328), which implicitly excludes them from your project. By overriding `CompilerGeneratedFilesOutputPath` you're opting out of the default behavior you desire.

**Reply by jasonmalinowski** [reply]
Indeed, emitting the generated files is something we generally expect you're going to put into an output directory, either because you're trying to debug a generator, or perhaps because you want to capture the output and store it with your other build artifacts. We don't generally expect you're emitting back into your source directory.

**Reply by papafe** [reply]
As @jasonmalinowski mentioned, my main reasons for having the generated files emitted to disk were two:
- Debugging, as sometimes the caching of the source generator in the UI made difficult to understand the effect of some changes. 
- Source control, in order to be able to keep track of how the source generator output changes over time
The default folder is probably a bad choice for both, as it's generally not part of source control, and hidden enough to be inspected. 

In the end it is a small change to exclude the generated files from compilation, but it was just a little surprising to me as a developer the first time 😄
